src/gui/render.cpp

```diff
@@ -527,364 +542,144 @@ void RENDER_SetSize(const uint16_t width, const uint16_t height,
+	force_no_pixel_doubling |= shader_info.settings.force_no_pixel_doubling;
+#endif
+
+	VGA_EnableVgaDoubleScanning(!force_vga_single_scan);
+	VGA_EnablePixelDoubling(!force_no_pixel_doubling);
 }
 
-static bool read_shader_source(const std::string& shader_path, std::string& source)
+bool RENDER_MaybeAutoSwitchShader([[maybe_unused]] const uint16_t canvas_width,
+                                  [[maybe_unused]] const uint16_t canvas_height,
+                                  [[maybe_unused]] const VideoMode& video_mode,
+                                  [[maybe_unused]] const bool reinit_render)
 {
-	// Start with the path as-is and then try from resources
-	const auto candidate_paths = {std_fs::path(shader_path),
-	                              std_fs::path(shader_path + ".glsl"),
-	                              GetResourcePath("glshaders", shader_path),
-	                              GetResourcePath("glshaders",
-	                                              shader_path + ".glsl")};
-
-	std::string s; // to be populated with the shader source
-	for (const auto& p : candidate_paths) {
-		if (read_shader(p, s)) {
-			break;
-		}
-	}
-
-	if (s.empty()) {
-		source.clear();
+#if C_OPENGL
+	// Currently, the init sequence is slightly different on Windows, macOS,
+	// and Linux. The first call to this function on Windows receives an
+	// uninitialised VideoMode param with width and height set to 0 which
+	// results in a crash. The proper fix is to make the init sequence 100%
+	// identical on all platforms, but in the interim this workaround will do.
+	if (canvas_width == 0 || canvas_height == 0 || video_mode.width == 0 ||
+	    video_mode.height == 0) {
 		return false;
 	}
 
-	if (first_shell) {
-		std::string pre_defs;
-		const size_t count = first_shell->GetEnvCount();
-		for (size_t i = 0; i < count; ++i) {
-			std::string env;
-			if (!first_shell->GetEnvNum(i, env)) {
-				continue;
-			}
-			if (env.compare(0, 9, "GLSHADER_") == 0) {
-				const auto brk = env.find('=');
-				if (brk == std::string::npos) {
-					continue;
-				}
-				env[brk] = ' ';
-				pre_defs += "#define " + env.substr(9) + '\n';
-			}
-		}
-		if (pre_defs.length()) {
-			// If "#version" occurs, it must be before anything,
-			// except comments and whitespace
-			auto pos = s.find("#version ");
+	get_shader_manager().NotifyRenderParametersChanged(canvas_width,
+	                                                   canvas_height,
+	                                                   video_mode);
 
-			if (pos != std::string::npos) {
-				pos = s.find('\n', pos + 9);
-			}
+	const auto new_shader_name = get_shader_manager().GetCurrentShaderInfo().name;
 
-			s.insert(pos, pre_defs);
-		}
-	}
-	if (s.empty()) {
-		source.clear();
-		LOG_ERR("RENDER: Failed to read shader source");
-		return false;
-	}
-	source = std::move(s);
-	assert(source.length());
-	return true;
-}
+	const auto changed_shader = (new_shader_name != render.current_shader_name);
```

Comment:
The `new_shader_name` is a bit confusing because we're asking the shader manager to get the "CurrentShaderInfo"'s name.

Should `CurrentShaderInfo` be `NewShaderInfo` instead? 

Or maybe `CurrentShaderInfo` is simply _more current_ (or newer) than the `render.current_shader_name`?

Reply:
Open to a better name, I'm not 100% happy with "current", but this is how the shader manager works; you feed it events that could potentially change the "current" (to be used) shader, then you query the "current" shader (that should be used now, in reflection of the incoming events). It's described here:

https://github.com/dosbox-staging/dosbox-staging/pull/2777/files#diff-506240744b4aec8b55f275b514845b582ae81f49c727bc4d0e3e54453a352f09R105-R121

So I think that comment described its function accurately, but yeah, open to suggestions. I don't think "GetCurrentShaderInfo" would be an improvement, maybe something else?

Reply:
Leaving as is for now.